Fill in client side SQL parameters.

vivekmenezes opened the pull request:
Process server side SQL parameters for INSERT statement.

petermattis commented on sql/driver/conn.go:
```diff
@@ -69,8 +70,33 @@ func (c *conn) Exec(stmt string, args []driver.Value) (driver.Result, error) {
 func (c *conn) Query(stmt string, args []driver.Value) (*rows, error) {
-	// TODO(vivek): Add the args to the Call.
-	return c.send(sqlwire.Call{Args: &sqlwire.Request{RequestHeader: sqlwire.RequestHeader{Session: c.session}, Sql: stmt}, Reply: &sqlwire.Response{}})
+	var params []sqlwire.Datum
```
Since we know the number of params, it is mildly more efficient to do:

```
  params := make([]sqlwire.Datum, 0, len(args))
```

petermattis commented on sql/driver/conn.go:
```diff
@@ -69,8 +70,33 @@ func (c *conn) Exec(stmt string, args []driver.Value) (driver.Result, error) {
+		case string:
+			param.StringVal = &value
+		case time.Time:
+			time, err := value.MarshalBinary()
```
How are we going to know on the server side that the bytes value as an encoded `time.Time`?

petermattis replied:
Might be better to encode the time as text in a standard format that our SQL executor expects.

bdarnell replied:
`time.Time.MarshalBinary` also uses an undocumented go-specific format; whatever we use here will need to be implementable from any language. For a textual format we should use ISO8601/RFC3339; we may also want a more compact binary format (adding a time-specific sub-message to `sqlwire.Datum` might be reasonable, or we could adopt and document go's binary format).

petermattis commented on sql/sqlserver/server.go:
```diff
@@ -686,7 +690,15 @@ func (s *Server) processInsertRows(node parser.InsertRows) (rows []sqlwire.Resul
 						vals = append(vals, sqlwire.Datum{StringVal: &tmp})
 					case parser.ValArg:
-						return rows, fmt.Errorf("TODO(pmattis): unsupported node: %T", val)
+						tmp := string(vt)
```
Add a `TODO` that filling in the placeholder values should really be done as a walk over the parse tree and not something specific to insert rows processing.

petermattis commented on sql/driver/conn.go:
```diff
@@ -69,8 +70,33 @@ func (c *conn) Exec(stmt string, args []driver.Value) (driver.Result, error) {
+		}
+		params = append(params, param)
+	}
+	return c.send(sqlwire.Call{Args: &sqlwire.Request{RequestHeader: sqlwire.RequestHeader{Session: c.session}, Sql: stmt, Params: params}, Reply: &sqlwire.Response{}})
```
Can you reformat this line? It is a bit unwieldy as a singe line.

petermattis commented on sql/parser/ast.go:
```diff
@@ -22,14 +22,14 @@ import (
-func Parse(sql string) (Statement, error) {
+// Parse parses the sql and returns a Statement and the number of
+// variables. The Statement is the AST representation of the query.
+func Parse(sql string) (Statement, int, error) {
```
This is fine for now, but I think the actual implementation should just walk the parse tree to determine a) the number of placeholders and b) to fill them in.

petermattis commented:
LGTM

bdarnell commented:
LGTM

vivekmenezes committed:
Fill in client side SQL parameters.
Process server side SQL parameters for INSERT statement and SELECT WHERE clause.

vivekmenezes commented on sql/server.go:
```diff
@@ -139,7 +198,10 @@ func (s *Server) exec(req driver.Request) (driver.Response, error) {
 	for _, stmt := range stmts {
-		// TODO(pmattis): Fill placeholders.
+		// Bind all the placeholder variables in the stmt to actual values.
+		if err = s.bindValues(stmt, parameters{params: req.Params}); err != nil {
```
We bind the values here. Another option is for us to bind them during expression evaluation. I suppose one advantage of doing it here is that it's done before the query execution plan is mapped out. Another possibility is to do it earlier while parsing.

A big problem of the current approach is that it needs to find all the expressions in the AST and it varies while we add new expressions to the AST.

petermattis replied:
I think binding the values here is correct. `parser.WalkExpr` is going to always walk the full expression tree. Rather than `Server.bindValues`, which doesn't actually need anything in `Server`, I think you should add `parser.WalkStmt` so that all of the binding occurs in the `parser` package.

vivekmenezes replied:
I've added an incomplete implementation of WalkStmt. It should be okay to check this in and follow up with another PR that implements it for all statements.

vivekmenezes commented:
I had to redo this change given that the code has changed a lot over the last two weeks.

petermattis commented on sql/server.go:
```diff
@@ -120,6 +120,65 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
+type parameters struct {
+	params []driver.Datum
+}
```
This could also be `type parameters []driver.Datum`.

vivekmenezes replied:
done

vivekmenezes committed:
Add WalkStmt. Apply FillArg on a Statement

tamird commented on sql/driver/conn.go:
```diff
@@ -62,8 +64,39 @@ func (c *conn) Exec(stmt string, args []driver.Value) (driver.Result, error) {
+	params := make([]Datum, 0, len(args))
+	for _, arg := range args {
+		if arg == nil {
+			return nil, errors.New("Passed in a nil parameter")
```
can you instantiate this error just once?

vivekmenezes replied:
I suppose so but it actually should never happen under a production scenario. It can happen when someone is playing around..

tamird replied:
are you saying it's actually impossible? we should panic if so

vivekmenezes replied:
The spec says that Value can be nil
https://golang.org/pkg/database/sql/driver/#Value

I'm leaving it as is.

petermattis replied:
I would guess that `nil` should be interpreted as `NULL`, not an error. Can probably test this with the postgres or mysql Go drivers. Or look at the source.

tamird replied:
@vivekmenezes ping

tamird commented on sql/driver/driver_test.go:
```diff
@@ -507,6 +506,59 @@ func TestSelectNoTable(t *testing.T) {
 		}
 	}
 }
+func TestSelectParameters(t *testing.T) {
```
i think this is a lot of boilerplate - how about just adding the two important lines to the existing select test?

vivekmenezes replied:
Yeah I'm inclined to do that. Let me do that.

vivekmenezes replied:
done

tamird commented on sql/parser/walk_test.go:
```diff
@@ -82,12 +85,11 @@ func TestFillArgs(t *testing.T) {
 		}
-		expr := q[0].(*Select).Exprs[0].(*NonStarExpr).Expr
-		expr, err = FillArgs(expr, d.args)
+		err = FillArgs(q[0], d.args)
```
can you use the shadowing form here and below? `if err := ...; err != nil {`

vivekmenezes replied:
done

tamird commented on sql/server.go:
```diff
@@ -139,7 +162,10 @@ func (s *Server) exec(req driver.Request) (driver.Response, error) {
 	for _, stmt := range stmts {
-		// TODO(pmattis): Fill placeholders.
+		// Bind all the placeholder variables in the stmt to actual values.
+		if err = parser.FillArgs(stmt, parameters(req.Params)); err != nil {
```
i think you want `err := ` here

vivekmenezes replied:
done

tamird commented:
LGTM

vivekmenezes committed:
Incorporate feedback

petermattis commented on sql/parser/walk.go:
```diff
@@ -162,8 +162,34 @@ func (v *argVisitor) Visit(expr Expr) Expr {
+		}
+		if stmt.Where != nil {
+			stmt.Where.Expr = WalkExpr(v, stmt.Where.Expr)
+		}
```
Add a `TODO` (or implement) the remaining portions of `parser.Select`. In particular, you're missing `stmt.Having`, `stmt.GroupBy`, `stmt.OrderBy` and `stmt.Limit`.

petermattis replied:
You also need to walk down into `Select.From` in order to handle the table expressions.

tamird replied:
@vivekmenezes ping

tamird replied:
disregard!

petermattis commented:
LGTM, though perhaps file an issue about the incompleteness of `WalkStmt`.

vivekmenezes commented:
Pete,

It's pretty incomplete and I'll just keep the current issue open until its
all done. I'm going to check this in and send you another PR with the rest
of the stuff.

Vivek

On Wed, Jul 29, 2015, 7:11 PM Peter Mattis <notifications@github.com> wrote:

> LGTM, though perhaps file an issue about the incompleteness of WalkStmt.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1659#issuecomment-126123515>
> .
>
